### PR TITLE
egw: escape vlan interface names in sysctl rpfilter path

### LINKED ISSUE #34217: Egress Gateway does not work if Egress Interface is a VLAN interface
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

higher than v1.15.7 and lower than v1.16.0

### What happened?

Since v1.15.6 Egress Gateway does not work anymore if the Egress GW interface is a VLAN interface in the format `ens2f0np0.201`

It seems this issue was introduced by commit https://github.com/cilium/cilium/commit/56e5133ccddafa48ebcef5fb7ff7da33f7ea9148

The "." in the interface name is interpreted by sysctl as another "directory" in the /proc/sys/net path. This results in the sysctl value `net.ipv4.conf.ens2f0np0.201.rp_filter`. As a result it tries to write to `/proc/sys/net/ipv4/conf/ens2f0np0/201/rp_filter`

### How can we reproduce the issue?

1. Install Cilium >= v1.15.6
2. Create CiliumEgressGatewayPolicy for pod

### Cilium Version

v1.15.6 to v1.16.0 (latest)

### Kernel Version

6.9.5-zabbly+ #debian12 SMP PREEMPT_DYNAMIC Mon Jun 17 04:03:53 UTC 2024 x86_64 GNU/Linux

### Kubernetes Version

v.1.29.5

### Regression

This worked in v1.15.5 or below.

### Sysdump

_No response_

### Relevant log output

```shell
FAIL queued &{net.ipv4.conf.ens2f0np0.201.rp_filter 2 false  Pending (1.4s ago)} for retry due to could not open the sysctl file /proc/sys/net/ipv4/conf/ens2f0np0/201/rp_filter: open /proc/sys/net/ipv4/conf/ens2f0np0/201/rp_filter: no such file or directory ! revision not equal
FAIL queued &{net.ipv4.conf.ens2f0np0.201.rp_filter 2 false  Pending (1.4s ago)} for retry due to could not open the sysctl file /proc/sys/net/ipv4/conf/ens2f0np0/201/rp_filter: open /proc/sys/net/ipv4/conf/ens2f0np0/201/rp_filter: no such file or directory ! revision not equal
```

### Anything else?

Workaround is to change interface name to a name without a dot (".") inside, e.g.
```shell
ip link add link ens2f0np0 name vlan201 type vlan id 201
```

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by henry-spanka**
This PR escapes "." meta character in sysctl path. VLAN interface names contain a "." (dot) which is interpreted by sysctl as new directory.

Fixes: #34217

```release-note
egw: Escape vlan interface names in sysctl rpfilter path
```

### CODE REVIEW

**Comment by liyihuang** [comment]
Any update on this PR? I ran into the same issue in my test environment and trying to see when this can be merged and backpacked to 1.15

**Comment by henry-spanka** [comment]
I fixed checks in tests and renamed second egress interface. Is this sufficient?

**Comment by julianwiedmann** [comment]
@henry-spanka just checking - did you test your patch locally? The [CI fail](https://github.com/cilium/cilium/actions/runs/10404992978/job/28814824223#step:17:358) for the test you added indicates to me that it's not working as expected, since the sysctl gets rejected [here](https://github.com/cilium/cilium/blob/fc9f58248b3ee5828c6ab305ab303cc736d14e81/pkg/datapath/linux/sysctl/sysctl.go#L283).

But with the test you added, you should be able to iterate more quickly now :)

**Comment by ysksuzuki** [comment]
I think we need to fix the sysctl package to be able to handle the case where the link name contains dots. Looks like we can't let the sysctl package to process  a path such as `/proc/sys/net/ipv4/conf/cilium_egw2.100` currently
https://github.com/cilium/cilium/blob/4219e5c3292612b7ecd221d39a683de1996ce71a/pkg/datapath/linux/sysctl/sysctl.go#L279

**Comment by henry-spanka** [comment]
Sort of. I'll take a look later and make sure tests pass ...

File: pkg/egressgateway/manager.go [hunk 1]
```diff
@@ -611,7 +611,7 @@ func (manager *Manager) relaxRPFilter() error {
 		if _, ok := ifSet[ifaceName]; !ok {
 			ifSet[ifaceName] = struct{}{}
 			sysSettings = append(sysSettings, tables.Sysctl{
-				Name:      fmt.Sprintf("net.ipv4.conf.%s.rp_filter", ifaceName),
+				Name:      fmt.Sprintf("net.ipv4.conf.%s.rp_filter", strings.ReplaceAll(ifaceName, ".", "/")),
```
Comment (dylandreimerink) [inline pkg/egressgateway/manager.go:614]: This doesn't actually fix the issue, when the sysctl package converts the remaining dots you end up with the same result.

**Review by dylandreimerink — CHANGES_REQUESTED** [review]
I suggest we fix this in the sysctl package by creating new versions of the existing functions which take a `[]string` instead of a `string` as `Name`. The sysctl can then join the slice without having to reserve any symbols as separator. This will allow for dots in the individual sections.

We should then also look over the rest of the codebase to see where we use interface names, its likely a lot of places, and start using the new functions there as well. (We should make an issue to track this, doesn't have to happen in this PR)

**Comment by dylandreimerink** [comment]
I made https://github.com/cilium/cilium/pull/34527 which fixes the root cause as described, including for all other callsites.

**Comment by joestringer** [comment]
@henry-spanka do you think you could take a look at #34527 and review or test it out to see if it addresses the same issue you're observing?

**Comment by joestringer** [comment]
> We should then also look over the rest of the codebase to see where we use interface names, its likely a lot of places, and start using the new functions there as well. (We should make an issue to track this, doesn't have to happen in this PR)

@dylandreimerink do you think you're likely to have more context about how to find those? I would welcome @henry-spanka to file the issue if they have the context, but as a first-time contributor that might be difficult.

**Comment by dylandreimerink** [comment]
> @dylandreimerink do you think you're likely to have more context about how to find those?

For https://github.com/cilium/cilium/pull/34527 I used the find all references functionality of gopls/vscode to find all usages, its a bit of work but doable.

**Comment by julianwiedmann** [comment]
https://github.com/cilium/cilium/pull/34527 resolved this for `v1.16` and `main`, while https://github.com/cilium/cilium/pull/34830 will bring the toleration for errors when setting RP filter into `v1.15`.

So I think the issue is addressed - I'll flip this PR to draft state so it falls off my review queue. @henry-spanka Please close the PR if you agree, thank you for bringing this problem up! :)

**Comment by henry-spanka** [comment]
Yes looks good. Thank you very much.

**Pull request closed without merging**